### DISCUSSION TITLE
Assistance required for connecting to corporate gitlab instance requiring x509 client certs.

### DISCUSSION THREAD

**Discussion by TauPan** [magit/forge#440] [Support] [answered]
Hi!

I've already posted this on [reddit](https://www.reddit.com/r/emacs/comments/r5rk7i/pkcs11_client_certificates_from_hardware_token/) but from glancing at the code it seems that forge simply does not support authentication via x509 client certs on a gitlab instance (which would not even require a token).

Would there be an (easy) way to subclass `forge-gitlab-repository` to support it?

**Comment by tarsius** [comment]
You could define a `forge-gitlab-x509-class` and would have to define new methods for `ghub--auth`, `ghub--username` and maybe others. It would probably better to modify the default methods eventually, but for early experiments separate methods should work.

**Reply by TauPan** [reply]
Thanks for the hints! I'll try my hand at this later. I've never worked with cl-classes in emacs (and I think I didn't do much of this in common lisp) so this might be fun.

**Reply by TauPan** [reply]
Hm, turns out this is harder for me than I can be bothered with at the moment. Sorry.

Edit: to be more specific: Since I've never worked with CLOS **and** this particular code I currently fail to see how to disable user authentication (since our gitlab **only** requires a client certificate). I'm particular confused that some methods seem to dispatch on symbols rather than types. But admittedly I've never even fully read Grahams' "On Lisp" ... it's sitting on my shelf and I remember reading the first few chapters and amiring the flexibility of the analytic approach to object orientation... Eh, sorry, too much information I guess.

I've tried overriding `forge--ghub-type-symbol` but either I return a new symbol `gitlab-x509` which then probably leads to me having to override all the methods, or I return `gitlab` which seems to inevitably leads to code being dispatched on that symbol which I never get to override.

This is my (probably very wrong) attempt:
~~~elisp
(defclass forge-gitlab-x509-repository (forge-gitlab-repository) nil)

(cl-defmethod ghub--auth ((forge (eql 'gitlab-x509)))
  nil)
(cl-defmethod ghub--username ((repo forge-gitlab-x509-repository))
  nil)

(defun forge--ghub-type-symbol (class)
  (cl-ecase class
    ;; This package does not define a `forge-gitlab-http-repository'
    ;; class, but we suggest at #9 that users define such a class if
    ;; they must connect to a Gitlab instance that uses http instead
    ;; of https.
    (forge-gitlab-x509-repository 'gitlab-x509)
    ((forge-gitlab-repository forge-gitlab-http-repository) 'gitlab)
    (forge-github-repository    'github)
    (forge-gitea-repository     'gitea)
    (forge-gogs-repository      'gogs)
    (forge-bitbucket-repository 'bitbucket)))
~~~

**Reply by Thaodan** [reply]
Did you try this again? For Git I just overrode `http.sslCert` with the path to my PKCS#11 object and that's it.
Defining something like `ghub--cert` as an option that could be forwarded could work.

**Reply by Thaodan** [reply]
If authsource is used the cert can be set for the connection and then used here.

**Reply by ghost** [reply]
I am trying to achieve something similar, with the exception of not needing to disable the token. So the connection is authorized by a reverse proxy with client certificates while the Gitlab authenticates with the token. Did anyone manage to do that, and would be kind enough to share how they did it? Or if you could guide me on where to start, I'd also appreciate that :smile:

**Reply by Thaodan** [reply]
I investigated the issue a bit deeper. The issue is that url doesn't have an option to forward more options to the `(open-network-stream)` call.
If that was possible one could try to find an matching cert for that host and then pass it to the network stream.

EDIT: Setting `network-stream-use-client-certificates should be enough`.

**Reply by tarsius** [reply]
Once you got it working, please post instructions for the benefit of others in the same boat.

**Reply by TauPan** [reply]
With a **bit** more recent forge (20220101.810 ... I didn't update in a while, but I made the initial attempt in 2021) it works like this:

 - Create a personal gitlab access token and put it into `~/.authinfo.gpg` as described in "File: ghub.info,  Node: Storing a Token"
 - Make sure `network-stream-use-client-certificates` is enabled as described in https://github.com/magit/forge/discussions/440#discussioncomment-4017740

That's it.

**Comment by TauPan** [comment, marked as answer]
So the correct answer as of January 2022 should be https://github.com/magit/forge/discussions/440#discussioncomment-4017740

No overriding of forge code necessary.
